# Movies Tonight Part 3: SQL DML

In [1]:
%load_ext sql
%sql sqlite:///Movies.db

'Connected: @Movies.db'

Open a git terminal 
Enter:
sqlite3 Movies.db
use .help for dot commands
%so it knows it's reading a csv%:
.mode csv
%import the dataset%
.IMPORT dataset.CSV DATASET
%show the tables that you have%
.table

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

In [9]:
%%sql
SELECT * FROM DATASET LIMIT 10;
PRAGMA table_info(Movies);

 * sqlite:///Movies.db
Done.
Done.


[(0, 'MovieID', 'INTEGER', 0, None, 1),
 (1, 'Rating', 'TEXT', 0, None, 0),
 (2, 'Title', 'TEXT', 1, None, 0)]

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [30]:
%%sql
DELETE FROM MOVIES;
INSERT INTO MOVIES (Title, Rating)
    SELECT DISTINCT MTitle, Rating FROM DATASET;
DELETE FROM THEATER;
INSERT INTO THEATER (Name, Location, Phone)
    SELECT DISTINCT TName, Location, Phone FROM DATASET;
DELETE FROM ARTIST;
INSERT INTO ARTIST (Name)
    SELECT DISTINCT CName FROM DATASET;
DELETE FROM SHOWTIME;
INSERT INTO SHOWTIME (Time, MovieID, TheaterID)
SELECT DISTINCT DATASET.ShowTime, MOVIES.MovieID, THEATER.TheaterID
FROM DATASET
    JOIN Movies ON (MOVIES.Title = DATASET.MTitle)
    JOIN Theater ON (THEATER.Name = DATASET.TName);
DELETE FROM CREDIT;
INSERT INTO CREDIT (MovieId, ArtistID, CCode)
SELECT DISTINCT MOVIES.MovieID, ARTIST.ArtistID, CCode
FROM DATASET
    JOIN Movies ON(MOVIES.Title = DATASET.MTitle)
    JOIN Artist ON(ARTIST.Name = DATASET.CName);

 * sqlite:///Movies.db
23 rows affected.
23 rows affected.
9 rows affected.
9 rows affected.
153 rows affected.
153 rows affected.
131 rows affected.
131 rows affected.
0 rows affected.
162 rows affected.


[]

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

In [19]:
%%sql
SELECT *
FROM MOVIES;
SELECT *
FROM THEATER;
SELECT *
FROM ARTIST;

 * sqlite:///Movies.db
Done.
Done.
Done.


[(1, 'Austin Pendleton', None, None),
 (2, 'Bebe Neuwirth', None, None),
 (3, 'Dianne Wiest', None, None),
 (4, 'Eli Wallach', None, None),
 (5, 'Kenny Kerr', None, None),
 (6, 'Lainie Kazan', None, None),
 (7, 'Tim Daly', None, None),
 (8, 'Whoopi Goldberg', None, None),
 (9, 'Donald Petrie', None, None),
 (10, 'Bernard Hill', None, None),
 (11, 'Brian McCardie', None, None),
 (12, 'Emily Mortimer', None, None),
 (13, 'Henry Cele', None, None),
 (14, 'John Kani', None, None),
 (15, 'Michael Douglas', None, None),
 (16, 'Tom Wilkinson', None, None),
 (17, 'Val Kilmer', None, None),
 (18, 'Stephen Hopkins', None, None),
 (19, 'Bill Pullman', None, None),
 (20, 'Bill Smitrovich', None, None),
 (21, 'Harvey Fierstein', None, None),
 (22, 'Jeff Goldblum', None, None),
 (23, 'Judd Hirsch', None, None),
 (24, 'Margaret Colin', None, None),
 (25, 'Mary McDonnell', None, None),
 (26, 'Randy Quaid', None, None),
 (27, 'Robert Loggia', None, None),
 (28, 'Vivica Fox', None, None),
 (29, 'Will Smith', None, None),
 (30, 'Roland Emmerich', None, None),
 (31, 'Emilio Estevez', None, None),
 (32, 'Heidi Kling', None, None),
 (33, 'Jeffrey Nordling', None, None),
 (34, 'Joss Ackland', None, None),
 (35, 'Rob Lieberman', None, None),
 (36, 'Greg Kinnear', None, None),
 (37, 'Hector Elizondo', None, None),
 (38, 'Jon Seda', None, None),
 (39, 'Laurie Metcalf', None, None),
 (40, 'Maria Pitillo', None, None),
 (41, 'Roscoe Lee Browne', None, None),
 (42, 'Tim Conway', None, None),
 (43, 'Garry Marshall', None, None),
 (44, 'Bette Midler', None, None),
 (45, 'Diane Keaton', None, None),
 (46, 'Goldie Hawn', None, None),
 (47, 'Hugh Wilson', None, None),
 (48, 'and Louise Fletcher', None, None),
 (49, 'Jon Lovitz', None, None),
 (50, 'Mekhi Phifer', None, None),
 (51, 'Tia Carrera', None, None),
 (52, 'Hart Bochner', None, None),
 (53, 'Bill Murray', None, None),
 (54, 'Janeane Garofalo', None, None),
 (55, 'Jeremy Piven', None, None),
 (56, 'Keith David', None, None),
 (57, 'Matthew McConaughey', None, None),
 (58, 'Pat Hingle', None, None),
 (59, 'Howard Franklin', None, None),
 (60, 'Barbra Streisand', None, None),
 (61, 'Jeff Bridges', None, None),
 (62, 'Lauren Bacall', None, None),
 (63, 'Pierce Brosnan', None, None),
 (64, 'Delroy Lindo', None, None),
 (65, 'Gary Sinese', None, None),
 (66, 'Lili Taylor', None, None),
 (67, 'Mel Gibson', None, None),
 (68, 'Rene Russo', None, None),
 (69, 'Ron Howard', None, None),
 (70, 'Billy Crudup', None, None),
 (71, 'Brad Pitt', None, None),
 (72, 'Brad Renfro', None, None),
 (73, 'Dustin Hoffman', None, None),
 (74, 'Jason Patric', None, None),
 (75, 'Joseph Perrino', None, None),
 (76, 'Kevin Bacon', None, None),
 (77, 'Minnie Driver', None, None),
 (78, 'Robert De Niro', None, None),
 (79, 'Ron Eldard', None, None),
 (80, 'Terry Kinney', None, None),
 (81, 'Vittorio Gassman', None, None),
 (82, 'Barry Levinson', None, None),
 (83, 'Michael Jordan', None, None),
 (84, 'Theresa Randle', None, None),
 (85, 'Wayne Knight', None, None),
 (86, 'Joe Pytka', None, None),
 (87, 'Joe Mantegna', None, None),
 (88, 'Kari Wuhrer', None, None),
 (89, 'Lucinda Jenney', None, None),
 (90, 'Michael Constantine', None, None),
 (91, 'Robert John Burke', None, None),
 (92, 'Tom Holland', None, None),
 (93, 'Aidan Quinn', None, None),
 (94, 'Alan Rickman', None, None),
 (95, 'Julia Roberts', None, None),
 (96, 'Liam Neeson', None, None),
 (97, 'Stephen Rea', None, None),
 (98, 'Neil Jordan', None, None),
 (99, 'Claire Danes', None, None),
 (100, 'Diane Venora', None, None),
 (101, 'John Leguizamo', None, None),
 (102, 'Leonardo DiCaprio', None, None),
 (103, 'Paul Rudd', None, None),
 (104, 'Paul Sorvino', None, None),
 (105, 'Baz Luhrmann', None, None),
 (106, 'Campbell Scott', None, None),
 (107, 'Ian Holm', None, None),
 (108, 'Isabella Rossellini', None, None),
 (109, 'Stanley Tucci', None, None),
 (110, 'Tony Shalhoub', None, None),
 (111, 'Anna Paquin', None, None),
 (112, 'Dana Delany', None, None),
 (113, 'Jef

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

 * sqlite:///Movies.db
(sqlite3.OperationalError) no such column: Movies.Title
[SQL: SELECT Movies.Title, Theater.Name;]
(Background on this error at: http://sqlalche.me/e/e3q8)


5. **Drop the `DATASET` table.**  
We don't need it anymore. 

In [34]:
%%sql 
DROP TABLE DATASET;

 * sqlite:///Movies.db
Done.


[]

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".

In [ ]:
Travis Fuller